cordapp-cpk/src/main/kotlin/net/corda/plugins/cpb/CpbPlugin.kt

```diff
@@ -0,0 +1,120 @@
+package net.corda.plugins.cpb
+
+import net.corda.plugins.cpk.*
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.artifacts.Dependency.ARCHIVES_CONFIGURATION
+import org.gradle.api.artifacts.ExternalDependency
+import org.gradle.api.artifacts.ModuleDependency
+import org.gradle.api.artifacts.ProjectDependency
+import org.gradle.api.tasks.bundling.Jar
+import org.gradle.api.tasks.bundling.ZipEntryCompression
+import java.nio.file.Files
+import java.util.jar.JarInputStream
+import java.util.zip.ZipInputStream
+
+@Suppress("UnstableApiUsage")
+class CpbPlugin : Plugin<Project> {
+
+    companion object {
+        private const val CPB_TASK_NAME = "cpb"
+        private const val CPB_CONFIGURATION_NAME = CPB_TASK_NAME
+        private const val CPB_PACKAGING_CONFIGURATION_NAME = "cpbPackaging"
+        private const val CPB_ARTIFACT_CLASSIFIER = "package"
+        private const val CPB_FILE_EXTENSION = CPB_TASK_NAME
+        private const val CPK_FILE_SUFFIX = ".$CPK_FILE_EXTENSION"
+        private const val JAR_FILE_SUFFIX = ".jar"
+        private val EXCLUDED_CPK_TYPES = setOf("corda-api")
+    }
+
+    override fun apply(project: Project) {
+        project.pluginManager.apply(CordappPlugin::class.java)
+        val allCordappsConfiguration = project.configurations.getByName(ALL_CORDAPPS_CONFIGURATION_NAME)
+
+        val cpbConfiguration = project.configurations.create(CPB_CONFIGURATION_NAME)
+            .extendsFrom(allCordappsConfiguration)
+            .setVisible(false)
+            .apply {
+                isCanBeConsumed = false
+                isCanBeResolved = false
+            }
+
+        val cpbPackaging = project.configurations.create(CPB_PACKAGING_CONFIGURATION_NAME)
+            .setTransitive(false)
+            .setVisible(false)
+            .withDependencies { dependencies ->
+                // Force Gradle to discover any transitive CPKs
+                // and include them in cpbConfiguration.
+                project.configurations.detachedConfiguration()
+                    .setTransitive(false)
+                    .setVisible(false)
+                    .extendsFrom(cpbConfiguration)
+                    .resolvedConfiguration
```

Comment:
if I understand correctly, this creates a temporary configuration that is then immediately garbage collected...

Reply:
why don't we resolve `cpbConfiguration` if that is the point of this bit of code?

Reply:
Because we are not interested in the resolved contents of the `cpb` configuration. The only reason to resolve anything at all is to force Gradle to execute the `withDependencies` handler on `allCordapps`. In theory, we could extend `allCordapps` instead of `cpb` and then resolve that instead. However, I suppose someone _could_ add a `withDependencies` handler to `cpb` at some point, in which case we'd need that to execute that handler too before iterating over `cpb.allDependencies`.

Reply:
but why is the `cpb` configuration not transitive as well? do we want to force developers to state all of their dependencies one by one as opposed to just declare their direct dependencies?

Reply:
oh, no sorry it's the `cpbPackaging` configuration that is not transitive, disregard my previous comment

Reply:
AFAIK the "transitive" flag only means something if the `Configuration` can also be resolved. The "transitive CPK" mechanism _does_ try to honour this flag too, although it is actually implemented by the `CordappDependencyCollector`... :male_detective:.

Reply:
I've noticed that even if it isn't clear to me why we've had to do that, resolving the configuration eagerly

Reply:
We need Gradle to execute all `withDependencies` handlers for `cpb` and `allCordapps` _before_ it executes this `withDependencies` handler for `cpbPackaging`, as otherwise `cpb.allDependencies` will not contain everything.

I don't see an obvious way of controlling the order in which Gradle executes these handler, other than [doing what Gradle says is required to invoke it](https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/Configuration.html#withDependencies-org.gradle.api.Action-).